src/util/src/algebra.rs

```diff
@@ -0,0 +1,582 @@
+    ///
+    /// This function returns `None` if the given order is not prime.
+    /// Checking the validity of inputs is expensive.
+    /// A field should therefore be constructed once and then reused as much as possible.
+    ///
+    /// Alternatively, one can use fixed, *trusted/tested* parameters with [`ScalarField::new_unchecked`].
+    pub fn new(order: BigUint, csprng: &mut Csprng) -> Option<Self> {
+        let f = ScalarField { q: order };
+        if f.is_valid(csprng) {
+            return Some(f);
+        }
+        None
+    }
+
+    /// Constructs a new scalar field from a given order without checking primality.
+    ///
+    /// This function *assumes* that the given order is prime.
+    pub fn new_unchecked(order: BigUint) -> Self {
+        ScalarField { q: order }
+    }
+
+    /// The function validates the given field by checking that the modulus is prime. The call is expensive.
+    pub fn is_valid(&self, csprng: &mut Csprng) -> bool {
+        is_prime(&self.q, csprng)
+    }
+
+    /// Returns one, the neutral element of multiplication, as a field element.
+    pub fn one() -> FieldElement {
+        FieldElement(BigUint::one())
+    }
+
+    /// Returns zero, the neutral element of addition, as a field element.
+    pub fn zero() -> FieldElement {
+        FieldElement(BigUint::zero())
+    }
+
+    /// Returns a random field element, i.e., a uniform random integer in `[0,q)` where `q` is the field order.
+    ///
+    /// The given `csprng` is assumed to be a secure randomness generator.
+    pub fn random_field_elem(&self, csprng: &mut Csprng) -> FieldElement {
+        FieldElement(csprng.next_biguint_lt(&self.q))
+    }
+
+    /// Returns the order `q` of the field
+    pub fn order(&self) -> BigUint {
+        self.q.clone()
+    }
+
+    /// Returns the length of the byte-array representation of field order `q`.
+    ///
+    /// For the standard parameter field this is `32`.
+    pub fn l_q(&self) -> usize {
+        (cnt_bits_repr(&self.q) + 7) / 8
+    }
+}
+
+/// An element of the multiplicative group `Z_p^r` as defined by [`Group`].
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct GroupElement(
+    #[serde(
+        serialize_with = "crate::biguint_serde::biguint_serialize",
+        deserialize_with = "crate::biguint_serde::biguint_deserialize"
+    )]
+    BigUint,
+);
+
+/// The group `Z_p^r`, a multiplicative subgroup of `Z_p`.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct Group {
+    /// Prime modulus `p`.
+    #[serde(
+        serialize_with = "crate::biguint_serde::biguint_serialize",
+        deserialize_with = "crate::biguint_serde::biguint_deserialize"
+    )]
+    p: BigUint,
+    /// Cofactor of `q` in `p − 1``.
+    #[serde(
+        serialize_with = "crate::biguint_serde::biguint_serialize",
+        deserialize_with = "crate::biguint_serde::biguint_deserialize"
+    )]
+    r: BigUint,
```

Comment:
I'm not sure that I understand why we store `r` here. Is it necessary?

Reply:
I see that it was already in the `FixedParameters` earlier. Maybe it's fine to have.

Reply:
We could only store `q` or `r` and compute the other on the fly. But both of them are used multiple times, so it seems cheaper to compute them once and store both of them.

Reply:
Decided to remove `r`, see discussion below.